Shell/Twee.cpp

```diff
@@ -0,0 +1,219 @@
+/*
+ * This file is part of the source code of the software program
+ * Vampire. It is protected by applicable
+ * copyright laws.
+ *
+ * This source code is distributed under the licence found here
+ * https://vprover.github.io/license.html
+ * and in the source directory
+ */
+
+/**
+ * Implements Twee's goal-directed preprocessing technique for equational reasoning.
+ * To quote "Twee: An Equational Theorem Prover" (Smallbone, 2021):
+ *
+ * ----- 
+ * Twee’s frontend can optionally transform the problem to make the prover more goal-directed.
+ * The transformation is simple, but strange.
+ * For every function term f(...) appearing in the goal, we introduce a fresh constant symbol a and add the axiom f(...) = a
+ * For example, if the goal is f(g(a), b) = h(c) we add the axioms f(g(a), b) = sF0, g(a) = sF1, and h(c) = sF2.
+ * Simplification will rewrite the first axiom to f(sF1, b) = sF0 and the goal to sF0 = sF2.
+ * -----
+ * 
+ * We take an intepretation of this for our purposes, NB:
+ *  - any ground term, in a clause derived from the conjecture, is transformed in this way
+ *  - introduced definitions are considered derived from the conjecture
+ */
+
+#include "Kernel/Clause.hpp"
+#include "Kernel/Problem.hpp"
+#include "Kernel/TermIterators.hpp"
+#include "Kernel/TermTransformer.hpp"
+#include "Kernel/Renaming.hpp"
+
+#include "Twee.hpp"
+
+using Kernel::Clause;
+using Kernel::Literal;
+using Kernel::Problem;
+using Kernel::Unit;
+using Kernel::UnitList;
+
+/*
+ * The actual worker for the twee trick.
+ # - evaluates conjucture literals bottom up (as orchestrated by Twee::apply)
+ # - and eagerly introduces new definitions for its subterms
+ # - already encountered subterms reuse older definitions
+ */
+class Definizator : public BottomUpTermTransformer {
+  public: // so that Twee::apply can directly access
+    // all the new definitions (as clauses) introduced along the way
+    UnitList* newUnits;
+
+    // accumlating the defining units as premises for the current "rewrite"
+    // (reset responsibility is with Twee::apply)
+    UnitList* premises;
+
+    // for each relevant term, cache the introduced symbol and the corresponding definition
+    DHMap<Term*,std::pair<unsigned,Clause*>> _cache;
+
+    Definizator(bool groundOnly) : newUnits(UnitList::empty()), _groundOnly(groundOnly) {}
+  private:
+    bool _groundOnly;
+
+    unsigned _typeArity;
+    TermStack _allVars; // including typeVars, which will come first, then termVars
+    TermStack _termVarSorts;
+
+    // a helper function to collect terms variables and their sorts
+    // all stored in the above private fields to be looked up by transformSubterm
+    void scanVars(Term* t) {
```

Comment:
I take it this isn't some variation on `SortHelper::collect*`?

Reply:
It's did not look like an exact match of any. Although the polymorphic part is "inspired" by @ibnyusuf 's work in Skolem.